pkg/cmd/drt-run/operations.go

```diff
@@ -0,0 +1,314 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"os/exec"
+	"os/signal"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/operations"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+func ctrlC(ctx context.Context, cancel context.CancelFunc) {
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt)
+	go func() {
+		<-sig
+		// Signal runners to stop.
+		cancel()
+	}()
+}
+
+// opsRegistry is a registry.Registry implementation which registers
+// operations for use in drt-run only.
+type opsRegistry struct {
+	ops []registry.OperationSpec
+}
+
+// MakeClusterSpec implements the registry.Registry interface.
+func (o *opsRegistry) MakeClusterSpec(nodeCount int, opts ...spec.Option) spec.ClusterSpec {
+	return spec.MakeClusterSpec(nodeCount, opts...)
+}
+
+// Add implements the registry.Registry interface.
+func (o *opsRegistry) Add(spec registry.TestSpec) {
+	panic("unimplemented")
+}
+
+// AddOperation implements the registry.Registry interface.
+func (o *opsRegistry) AddOperation(spec registry.OperationSpec) {
+	o.ops = append(o.ops, spec)
+}
+
+// PromFactory implements the registry.Registry interface.
+func (o *opsRegistry) PromFactory() promauto.Factory {
+	panic("unimplemented")
+}
+
+var _ registry.Registry = &opsRegistry{}
+
+// opsRunner holds all state for all `parallel` operation workers
+// being run at the moment.
+type opsRunner struct {
+	specs       [][]registry.OperationSpec
+	config      config
+	l           *logger.Logger
+	registry    registry.Registry
+	parallelism int
+	seed        int64
+	eventLogger *eventLogger
+
+	mu struct {
+		syncutil.Mutex
+
+		completed         sync.Cond
+		lastRun           map[string]time.Time
+		runningOperations []string
+	}
+}
+
+// pickOperation picks one operation to run, that hasn't been run for at least `cadence`
+// time (specified in the YAML config).
+func (r *opsRunner) pickOperation(ctx context.Context, rng *rand.Rand) *registry.OperationSpec {
+	for {
+		if err := ctx.Err(); err != nil {
+			return nil
+		}
+		setIdx := rng.Intn(len(r.specs))
+		opSpecsSet := r.specs[setIdx]
+		opSpec := &opSpecsSet[rng.Intn(len(opSpecsSet))]
+		r.mu.Lock()
+		lastRun := r.mu.lastRun[opSpec.Name]
+		eligibleForNextRun := lastRun.Add(r.config.Operations.Sets[setIdx].Cadence)
```

Comment:
Following the logic here I think we could end up in scenarios where an operation of the same kind starts and overlaps with itself. It also brings up another question as to if we should have two exclusions options: One for not running concurrently, and another for supporting concurrency, but not with itself. The latter I think is problematic if an operation is waiting on clean-up (like disk stall or node kill), this logic could start that same operation again, and we'd be in a bad state, I think?

Reply:
I've made this change and updated the CanRunConcurrently type. PTAL!